test/server/cards/01_SOR/units/GuardianOfTheWhills.spec.ts

```diff
@@ -0,0 +1,143 @@
+                context.player1.clickCard(context.mandalorianArmor);
+                expect(context.player1).toBeAbleToSelectExactly([context.moistureFarmer]);
+                context.player1.clickCard(context.moistureFarmer);
+                expect(context.guardianOfTheWhills).toHaveExactUpgradeNames([context.entrenched.internalName, context.jediLightsaber.internalName]);
+                expect(context.moistureFarmer).toHaveExactUpgradeNames([context.mandalorianArmor.internalName]);
+                expect(context.player1.exhaustedResourceCount).toBe(7);
+
+                context.player2.passAction();
+
+                context.player1.clickCard(context.vambraceGrappleshot);
+                expect(context.player1).toBeAbleToSelectExactly([context.guardianOfTheWhills, context.moistureFarmer]);
+                context.player1.clickCard(context.guardianOfTheWhills);
+                expect(context.guardianOfTheWhills).toHaveExactUpgradeNames([context.entrenched.internalName, context.jediLightsaber.internalName, context.vambraceGrappleshot.internalName]);
+                expect(context.player1.exhaustedResourceCount).toBe(9); // shouldn't get a bonus (despite parent card changing for mandalorian armor)
+
+                // Test having an upgrade get removed and added back tracks correctly
+                context.moveToNextActionPhase();
+                context.player1.clickCard(context.protector);
+                expect(context.player1).toBeAbleToSelectExactly([context.guardianOfTheWhills, context.moistureFarmer, context.systemPatrolCraft, context.survivorsGauntlet]);
+                context.player1.clickCard(context.guardianOfTheWhills);
+                expect(context.player1.exhaustedResourceCount).toBe(2); // full cost is 3 given the double aspect, so 2 with discount
+
+                context.player2.clickCard(context.confiscate);
+                context.player2.clickCard(context.protector);
+                expect(context.protector).toBeInZone('discard');
+
+                context.player1.clickCard(context.razorCrestReliableGunship);
+                expect(context.player1).toBeAbleToSelectExactly([context.jetpack, context.devotion, context.electrostaff, context.protector]);
+                expect(context.player1).toHavePassAbilityButton();
+                context.player1.clickCard(context.protector);
+                expect(context.player1.exhaustedResourceCount).toBe(6);
+
+                context.player2.passAction();
+
+                context.player1.clickCard(context.protector);
+                expect(context.player1).toBeAbleToSelectExactly([context.guardianOfTheWhills, context.moistureFarmer, context.systemPatrolCraft, context.survivorsGauntlet, context.razorCrestReliableGunship]);
+                context.player1.clickCard(context.guardianOfTheWhills);
+                expect(context.player1.exhaustedResourceCount).toBe(9); // full cost is 3
```

Comment:
Could we add one more test where the upgrade is only playable with the discount (i.e., holding a 2 cost upgrade but only 1 ready resource) to confirm that the discount is being correctly counted when determining whether a card is legally playable. Can use `context.player1.exhaustResources()` to do it

Reply:
Starting this change in another branch as its far more involved.  We need to handle the fact that payForCost needs to check targets, and checkTarget needs to check payForCost.  Not sure how we'll cleanly do this